### DISCUSSION TITLE
Is it possible to read all the files' name in a directory using GitHub Cli?

### DISCUSSION THREAD

**Discussion by mendax1234** [community/community#57942] [Actions] [answered]
### Select Topic Area

Question

### Body

Is it possible to read all the files' name in a directory using GitHub Cli? If so, can I store the name into an array?

**Comment by airtower-luna** [comment, marked as answer]
You can use the [get repository content](https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content) API. If the `PATH` part of the URL is a directory, you'll get JSON describing the files & subdirectories in that directory.

Then process the JSON as you like. If you just want a list with one name per line, the `--jq` option of `gh` might be useful, e.g.

```
gh api --jq '.[].name' /repos/airtower-luna/luna/contents/src
```

will list the names of all the files in a directory in one of my repositories.

**Reply by mendax1234** [reply]
> You can use the [get repository content](https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content) API. If the `PATH` part of the URL is a directory, you'll get JSON describing the files & subdirectories in that directory.
> 
> Then process the JSON as you like. If you just want a list with one name per line, the `--jq` option of `gh` might be useful, e.g.
> 
> ```
> gh api --jq '.[].name' /repos/airtower-luna/luna/contents/src
> ```
> 
> will list the names of all the files in a directory in one of my repositories.

Hi @airtower-luna . I'd like to know how can l put the outcome of this command `gh api --jq '.[].name' /repos/airtower-luna/luna/contents/src` into array. And how to get an array's length in Github Action

**Reply by airtower-luna** [reply]
That depends on the programming language you're using, what I posted is simply a shell command (using [GitHub CLI](https://cli.github.com/manual/)) which you could use in a `run` step, and then process the result by normal shell means (none of this is Actions-specific syntax). For example you could count the number of lines using `wc`:

```
gh api --jq '.[].name' /repos/airtower-luna/luna/contents/src | wc -l
```

If you wanted to use Python it'd probably be best to avoid `gh` entirely and instead `import requests` and call the API that way. :wink:

**Reply by mendax1234** [reply]
> That depends on the programming language you're using, what I posted is simply a shell command (using [GitHub CLI](https://cli.github.com/manual/)) which you could use in a `run` step, and then process the result by normal shell means (none of this is Actions-specific syntax). For example you could count the number of lines using `wc`:
> 
> ```
> gh api --jq '.[].name' /repos/airtower-luna/luna/contents/src | wc -l
> ```
> 
> If you wanted to use Python it'd probably be best to avoid `gh` entirely and instead `import requests` and call the API that way. 😉

Thanks! @airtower-luna If l want to use each of the name l get, it seems that it is stored in the .[].name, so can l use .[].name in the shell command?

Besides, how can l store the number of the line into a variable using the shell command?

**Reply by airtower-luna** [reply]
`.[].name` is not shell syntax, it's template syntax like `jq` would use. The `--jq` option of `gh` is simply a convenience option to format data from the JSON the API call returns in a way you want. I highly recommend you look at the JSON you get *without* the `--jq '.[].name'` option to understand what's going on. :wink: 

> Besides, how can l store the number of the line into a variable using the shell command?

Simply capture the output in an environment variable (again nothing Actions-specific, just standard Unix shell):

```sh
lines=$(gh api --jq '.[].name' /repos/airtower-luna/luna/contents/src | wc -l)
```

If you want to use the value in later steps, you can then [set it as an environment variable or output for Actions](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#environment-files) (that part is Actions-specific).

**Reply by mendax1234** [reply]
> `.[].name` is not shell syntax, it's template syntax like `jq` would use. The `--jq` option of `gh` is simply a convenience option to format data from the JSON the API call returns in a way you want. I highly recommend you look at the JSON you get _without_ the `--jq '.[].name'` option to understand what's going on. 😉

Got it. So, if l want to store each name of the file in the directory into a variable like an array, what should l do? Thanks

Besides, l also meet with some strange problems. After l have used this command `lines=$(gh api --jq '.[].name' /repos/airtower-luna/luna/contents/src | wc -l)` and use `echo "$lines"` to check the output, l find that the first output is 4 but the second output is 16. That's strange. Can you help me with it ? Thanks!

**Reply by airtower-luna** [reply]
> So, if l want to store each name of the file in the directory into a variable like an array, what should l do?

I can only repeat that it depends on the language you're using. On the shell you'd use `var=($(gh ...))` to interpret the `gh` output as a whitespace-separated list and make an array from that (might break if your filenames contain spaces or other special characters). What's in a `run` step is just a shell script, unless you set `shell` to some other interpreter for that step. :wink: 

You *can't* put an array into an Actions environment variable or output, the closest you can do is to write JSON to it (or some other serialization).

> Besides, l also meet with some strange problems. After l have used this command lines=$(gh api --jq '.[].name' /repos/airtower-luna/luna/contents/src | wc -l) and use echo "$lines" to check the output, l find that the first output is 4 but the second output is 16. That's strange. Can you help me with it ?

Very strange, but impossible to answer without seeing the actual log & workflow. `wc -l` should produce exactly one number. :sweat_smile:
